src/geometry/CTiglBSplineAlgorithms.cpp

```diff
@@ -314,7 +314,138 @@ namespace
     {
         return array2GetRow<TColgp_Array2OfPnt, TColgp_HArray1OfPnt, Handle_TColgp_HArray1OfPnt>(matrix, rowIndex);
     }
-    		
+
+    // Some helper functions for CTiglBSplineAlgorithms::reparameterizePiecewiseLinear ...
+    void knotRefinementForReparam(const Handle(Geom_BSplineCurve) curve, std::vector<double> const& paramsOld)
+    {
+        // This function matches the first step of the reparameterization algorithm in The NURBS Book (2nd edition), p. 251
+
+        Standard_Integer nbKnots = curve->NbKnots();
+
+        //Start with increasing the multiplicity up to p
+        Standard_Integer degree = curve->Degree();
+
+        // Pass first and last index of knots (to increase all multiplicities until mult >= degree)
+        curve->IncreaseMultiplicity(1, nbKnots, degree);
+
+        // Add all knots u_i=f(s_i)
+        // Since we assume a piecewise linear reparameterization function f, the knots of f are just the new parameters
+        // Hence, evaluating f at one knot gives the corresponding old parameter
+        TColStd_Array1OfReal knotsParams(1, paramsOld.size());
+        TColStd_Array1OfInteger multsParams(1, paramsOld.size());
+        for (int paramIdx = 0; paramIdx < paramsOld.size(); paramIdx++) {
+            knotsParams.SetValue(paramIdx+1, paramsOld[paramIdx]);
+            multsParams.SetValue(paramIdx+1, degree);
+        }
+        curve->InsertKnots(knotsParams, multsParams);
+    }
+
+    double calcReparamfctInv(std::vector<double> const& paramsOld, std::vector<double> const& paramsNew, double u)
+    {
+        // Calc inverse of reparametrization function s=g(u)=f⁻¹(s)
+        // We set up BSpline of degree 1 -> g is piecewise linear
+        // In the interval [p_i,p_{i+1}], g(u)=m*u+n
+        // p_i and p_{i+1} are old parameters that need to be determined in the first place
+
+        // u needs to be in range of paramsOld
+        if (u < paramsOld[0] || paramsOld[paramsOld.size()-1] < u)
+            throw tigl::CTiglError("Argument of reparameterization function out of range in calcReparamfctInv");
+
+        // First: Find interval [p_i,p_{i+1}] to determine boundaries
+        // a_x = lowerBound, b_x = upperBound
+
+        double a_x, a_y, b_x, b_y;
+        auto it = std::find_if(
+            paramsOld.begin(),
+            paramsOld.end(),
+            [&u](double v){ return u <= v; }
+        );
+
+        int paramsIdx;
+        // Catch case u=0, since distance would be 0 which causes in illegal vector access with subtraction of 1
+        if (*it == 0)
+            paramsIdx = 1;
+        else
+            paramsIdx = std::distance(paramsOld.begin(), it);
```

Comment:
But this only works, if `paramsOld[0] == 0.`, right? What if `paramsOld[0] = -1.`? We should check if we get the first index, e.g. because `u == paramsOld[0]`. Then we should increment the index counter, which is the index of the parameter value of the right interval border.

```suggestion
        int paramsIdx = std::distance(paramsOld.begin(), it);;
        // Catch case u=0, since distance would be 0 which causes in illegal vector access with subtraction of 1
        if (paramsIdx  == 0) {
            ++paramsIdx;
        }
```